dlc-manager/src/chain_monitor.rs

```diff
@@ -108,41 +107,91 @@ impl ChainMonitor {
-                    let info = self.watched_tx.get(txid);
-                    if info.is_some() {
-                        return info;
-                    }
-                }
-                None
-            });
-            if let Some(channel_info) = channel_info {
-                res.push((tx.clone(), channel_info.clone()));
+        for (txid, state) in self.watched_tx.iter_mut() {
+            if let Some(tx) = block.txdata.iter().find(|tx| tx.txid() == *txid) {
+                state.confirm(tx.clone());
             }
         }
 
-        res
+        self.last_height += 1;
+    }
+
+    /// All the currently watched transactions which have been confirmed.
+    pub(crate) fn confirmed_txs(&self) -> Vec<(Transaction, ChannelInfo)> {
+        self.watched_tx
+            .iter()
+            .filter_map(|(_, state)| match state {
+                WatchState::Registered { .. } => None,
+                WatchState::Confirmed {
+                    channel_info,
+                    transaction,
+                } => Some((transaction.clone(), *channel_info)),
+            })
+            .collect()
     }
+}
 
-    /// To be safe this is a separate function from process block to make sure updates are
-    /// saved before we update the state. It is better to re-process a block than not
-    /// process it at all.
-    pub(crate) fn increment_height(&mut self, last_block_hash: &BlockHash) {
-        self.last_height += 1;
-        self.last_block_hashes.push(*last_block_hash);
-        if self.last_block_hashes.len() > NB_SAVED_BLOCK_HASHES {
-            self.last_block_hashes.remove(0);
+/// The state of a watched transaction.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum WatchState {
+    /// The transaction has been registered but we are not aware of
+    /// any confirmations.
+    Registered { channel_info: ChannelInfo },
+    /// The transaction has received at least one confirmation.
+    Confirmed {
+        channel_info: ChannelInfo,
+        transaction: Transaction,
+    },
+}
+
+impl_dlc_writeable_enum!(
+    WatchState,;
+    (0, Registered, {(channel_info, writeable)}),
+    (1, Confirmed, {(channel_info, writeable), (transaction, writeable)});;
+);
+
+impl WatchState {
+    fn new(channel_info: ChannelInfo) -> Self {
+        Self::Registered { channel_info }
+    }
+
+    fn confirm(&mut self, transaction: Transaction) {
+        match self {
+            WatchState::Registered { ref channel_info } => {
+                log::info!(
+                    "Transaction {} confirmed: {channel_info:?}",
+                    transaction.txid()
+                );
+
+                *self = WatchState::Confirmed {
+                    channel_info: *channel_info,
+                    transaction,
+                }
+            }
+            WatchState::Confirmed {
+                channel_info,
+                transaction,
+            } => {
+                log::warn!(
```

Comment:
nit: i'd probably put that at the error level unless there is some cases where this should happen without it being an error?

Reply:
Fair enough, will do!